WIP: handling viterbi breaks as multiple sequences

kodonnell opened the pull request:
#### Objective

If the viterbi sequence breaks (e.g. no candidates found for a given GPX entry, or no transitions between candidates) then instead of throwing an exception, split into a new sequence. As far as I'm aware there's currently no easy way for the user to do this manually, let alone easily.

#### Questions

- is this sensible behaviour for the user?
- should we return a single MapMatch result (with a single time/distance) - as it implemented in this first commit - or multiple MapMatch results (each within their own time/distance). I prefer the latter (with a reason why each break occurred).
- specific comments below
- anything else?

Cc @karussell @stefanholder

kodonnell commented on matching-core/src/main/java/com/graphhopper/matching/MapMatching.java:
```diff
@@ -166,6 +166,8 @@ public MatchResult doWork(List<GPXEntry> gpxList) {
         // Compute all candidates first.
         // TODO: Generate candidates on-the-fly within computeViterbiSequence() if this does not
         // degrade performance.
+        // @kodonnell: it's not currently possible as that'd mean calling queryGraph.lookup
```
@karussell / @stefanholder - happy for me to remove the above TODO?

karussell replied:
@stefanholder I probably do not understand this yet deep enough, but why do you think this would mean calling queryGraph.lookup multiple times?

Furthermore: creating a QueryGraph is cheap so we could call it then multiple times but the resulting QueryResults are independent.

kodonnell replied:
Not sure if you meant me - but I think it requires calling queryGraph.lookup multiple times because we're wanting to get the candidates at each step. E.g. first step: get candidates, lookup, do viterbi next step; second step: get candidates, lookup, do viterbi next step, ... 

If we create multiple queryGraphs, then we'll get duplicate virtual edge IDs, etc., which might be a problem.

stefanholder replied:
I think that looking up all candidates at once might be a problem for map matching very long GPS traces, e.g. from Munich to Hamburg. But maybe it's better to create an issue for this than having this TODO in the code.

kodonnell replied:
Good point @stefanholder - hopefully #90 will suffice.

kodonnell commented on matching-core/src/main/java/com/graphhopper/matching/MapMatching.java:
```diff
@@ -229,52 +231,77 @@ public MatchResult doWork(List<GPXEntry> gpxList) {
+                // recent step does not get added i.e. 'computeMostLikelySequence' returns the most
+                // likely sequence without this (breaking) step added. Hence we can use it to start
+                // the next one:
+                // TODO: check the above is true
```
@stefanholder - I think this is true, but can you confirm? I guess I could check for it to some degree ...

stefanholder replied:
Yes, your comment is true.

kodonnell commented on matching-core/src/main/java/com/graphhopper/matching/MapMatching.java:
```diff
@@ -301,70 +328,92 @@ private void computeTransitionProbabilities(TimeStep<GPXExtension, GPXEntry, Pat
                             .transitionLogProbability(path.getDistance(), linearDistance, timeDiff);
                     timeStep.addTransitionLogProbability(from, to, transitionLogProbability);
+                } else {
+                    // TODO: can we remove maxVisitedNodes completely and just set to infinity?
```
@karussell - what is the purpose of having maxVisitedNodes in map-matching? In theory we can more easily prevent large graph traversals by limiting linear distance ... and then the user doesn't need to worry about configuring this option (and we have a 'physical' reason for throwing an error or finding no candidate - there were no transitions found within a given radius).

karussell replied:
> If I want a map match, there's nothing I can do except increase maxVisitedNodes

If you have full control of the library&server then just use a maximum value, but defaults should be safe so that not a single request could occupy one CPU for dozens of seconds.

> As an aside - the default GH behaviour is presumably to have no limit on maxVisitedNodes?

No. We have a max nodes limits too for non-CH routes. But the limits are much bigger because the number of points are typically a lot less.

kodonnell replied:
OK. I'll leave it throwing an error for now. In future, we could maybe think of adding e.g. `matchResult.isFound()` similar to `path.isFound()`.

kodonnell commented on matching-core/src/main/java/com/graphhopper/matching/MapMatching.java:
```diff
@@ -301,70 +328,92 @@ private void computeTransitionProbabilities(TimeStep<GPXExtension, GPXEntry, Pat
+                    }
+                    // TODO: can we somewhere record that this route failed? Currently all viterbi
+                    // knows is that there's no transition possible, but not why. This is useful
+                    // for e.g. explaining why a sequence broke.
```
If we split into multiple MatchResult (one per each sequence) then maybe we could add a 'sequenceStartReason' property? For the first sequence it will be 'first sequence' and for the rest it might be 'no candidates found' or 'no transitions found' etc. @karussell / @stefanholder ?

karussell replied:
Sounds good to me, but I'm not sure if this is helpful for the user?

kodonnell replied:
Me neither. My thinking was that breaking into sequences is better than throwing an exception. However, I then thought that the user may want to know why the sequence broke. And it might be useful for diagnostics etc.

karussell replied:
Yes, I agree. Having separated sequences is the way to go IMO. And if user requests specific reasons we can also fine tune this later on.

kodonnell replied:
Good point - will leave as a TODO for now.

kodonnell commented on matching-core/src/main/java/com/graphhopper/matching/MapMatching.java:
```diff
@@ -301,70 +328,92 @@ private void computeTransitionProbabilities(TimeStep<GPXExtension, GPXEntry, Pat
+    private MatchResult computeMatchedEdges(
+            List<List<SequenceState<GPXExtension, GPXEntry, Path>>> sequences,
+            Map<String, EdgeIteratorState> virtualEdgesMap) {
+        // TODO: remove gpx extensions and just add time at start/end of edge.
```
@karussell - can I do this? I'm not sure why else we save all the gpxExtensions to the given edge ..

kodonnell replied:
Sorry - I want to remove all GPX extensions from edge matches (and hence match result), and add only the to/from time to the edge (which will be inferred from the GPX points and route times, etc.).

karussell replied:
The idea is that in the future it should (correctly) associate all original GPXEntry's to one edge. See e.g. https://discuss.graphhopper.com/t/map-matchresult-to-gpxentry/977

kodonnell replied:
Oh, that should (hopefully) be easy - I need to find the matching GPX entry to infer the start/end time anyway. Separate PR?

karussell replied:
> Separate PR?

Would be welcome :) ! But you would need to use some additional data and so we keep the GPXExtensions in the EdgeMatch?

kodonnell replied:
Not sure - will discuss in the (hopefully soon) PR.

stefanholder commented:
>is this sensible behaviour for the user?

Yes, this makes sense.

>should we return a single MapMatch result (with a single time/distance) - as it implemented in this first commit - or multiple MapMatch results (each within their own time/distance). I prefer the latter (with a reason why each break occurred).

I think that `MatchResult` should contain the matched GPS position for each original GPS position, as discussed [above](https://github.com/graphhopper/map-matching/pull/87#discussion_r92029486). Then each such entry could also contain a field indicating if/why an HMM break occurred. I think this is easier to handle by the user than multiple map match results. The user might not even be interested in HMM breaks and getting a single MapMatch result is then easier to handle.

kodonnell commented:
> I think that MatchResult should contain the matched GPS position for each original GPS position, as discussed above. Then each such entry could also contain a field indicating if/why an HMM break occurred.

Ah, that's a good idea. So, to clarify, we basically want one MapMatch result which includes the final route, as well as all the original GPX entries. Each entry can be inspected to know:

- the edge (of the route) that it is on (if any)
- some sort of status flag - 'start of sequence', 'single candidate', 'not used in map matching as too close to previous candidate', etc.

I'd suggest we also add a more general status flag to the MapMatch result e.g. 'contains sequence breaks', so users can easily check for this. And maybe we can warn the user too. (Otherwise if it's used blindly, we may end up with lots of issues submitted asking why there is a 'break' in the matched route!)

Anyway, with this, we should be able to describe all situations. However, as you say, most users will probably only be interested in the final route.

@stefanholder - does that sound good to you?

stefanholder commented:
>@stefanholder - does that sound good to you?

Yes, sounds good.

I would put it like this: `MatchResult` should contains the final route as well as a list of match entries.

Each match entry contains:
- the original GPX entry
- the map matched lat/lon
- the edge (of the route) that it is on (if any) and the offset from the start of the edge (or alternatively the fraction as offset divided by edge length)
- HMM break status: 'no break', 'no candidates', 'no transitions to this entry'
- 'used for map matching' or 'not used in map matching as too close to previous candidate'

kodonnell commented:
> I would put it like this:

Great, sounds good. I'll look over your other PR first, and then have a play with all of the comments above.

kodonnell commented:
You can ignore the code for now (it's still unfinished), but I'd like some feedback on progress thus far. Specifically:

1. I realised I didn't want to directly add in the GPX entries not used for map matching. Why? Because if the user e.g. loops through *all* of the GPX entires (including those not used for map-matching) they resulting sequence may look a lot more erratic and behave unexpectedly. (E.g. if I go from A -> B, when I add in the ignored points, it might be 'go toward B, then back toward A, then toward B ...'.) I think it makes more sense to think of clustering - i.e. points that are close to each other are the 'same'. For this reason, I do not snap the GPX entries (that aren't used) to the final route - but merely associate them with their 'cluster'. This means we avoid the erratic nature, but also that we can find where any of the original GPX entries is in the final route (up to the GPS error).
    - an alternative option is to leave it as is ('ignored' GPX entries are ignored from the result), but allow the user to change the filtering to prevent e.g. any filtering at all. This is simpler, but will slow down the matching.
2. I think the `MatchSequence` does most of what @stefanholder suggested [above](https://github.com/graphhopper/map-matching/pull/87#issuecomment-266981527) (though maybe not tidily yet). I provide a `MatchResult.originalGPXMapping` which allows the user to map (by index) a given original GPX entry to the matched one - and this contains information like the matched edge, how far along, the snapped point, the to/from time, etc. Alternatively one can traverse the matches directly.
    - a tidier (API and code) approach would be to add properties directly to the original GPXEntry. It means not using the GH GPXEntry (just extending it) but this probably isn't a big deal.
3. when we have sequence breaks, should I fill the break with an 'empty' sequence break, to keep the times contiguous? I.e. that way a user should always find a sequence at a given time.

The main changes remaining are 

- refactoring to get it nice and clean (I'll wait a few days and come back with a fresh view)
- make sure the to/from times are contiguous.
- trivial-ish refactoring of tests, etc., and add some new tests (maybe).
- refactoring the web GUI to handle sequences.

stefanholder commented:
>an alternative option is to leave it as is ('ignored' GPX entries are ignored from the result), but allow the user to change the filtering to prevent e.g. any filtering at all. This is simpler, but will slow down the matching.

I would prefer this alternative because it's simpler and I think the user should have the option to disable the filtering.

>a tidier (API and code) approach would be to add properties directly to the original GPXEntry

I think it's better to have a separate class (`GPXMapping`) for the matched GPX entry.

>when we have sequence breaks, should I fill the break with an 'empty' sequence break, to keep the times contiguous?

If there are no candidates for a GPX entry then `GPXMapping.matchEntry` should be null (not sure if this is what you mean).

kodonnell commented:
Thanks @stefanholder - I will follow your first two preferences. 

> If there are no candidates for a GPX entry then GPXMapping.matchEntry should be null (not sure if this is what you mean).

More if e.g. we have a sequence A->C then an HMM break so start a new sequence D->F. In this case, there's no (explicit) information about what happened between C and D - we could leave it to the user to interpret, or I could add an 'unknown'/'empty' sequence C->D which really only contains the information about the start/end points (including time) but e.g. the route between them will be 'unknown'. Does that make sense?

kodonnell commented:
To keep PR from getting even bigger, I created #92 for allowing customisable filtering. Otherwise, this PR is stabilising somewhat. I've renamed a lot of the entities - the names were misleading, and were confusing me (let alone new developers). E.g. GPXExtension -> Candidate, TimeStep -> ViterbiMatchEntry. I've updated the list above with some more tasks remaining too.

> should I fill the break with an 'empty' sequence break

For now I'll leave this out - it's non-essential, and isn't too hard to add later if it's desired.

stefanholder commented:
>More if e.g. we have a sequence A->C then an HMM break so start a new sequence D->F.

If an HMM break happens because there is no transition between C and D then we can set the HMM break status for the `GPXMapping` of D to 'no transitions to this entry', which is quite simple to do.

kodonnell commented:
> If an HMM break happens because there is no transition between C and D then we can set the HMM break status for the GPXMapping of D to 'no transitions to this entry', which is quite simple to do.

That's already done. To rephrase another way - say time at C is 10am and time at D is 11am. If the user currently goes to find 'what was the state at 10:30am', they could get confused, as they won't find anything (due to the sequence break). If we added a 'between break' sequence C->D, then they'd find that at 10:30am it was in an between-break state.  In other words, should we explicitly or implicitly identify the presence of a sequence break? Put another way, each sequence should be contiguous in time, but the sequence of sequences is not, and maybe it should be.

kodonnell commented on matching-core/src/main/java/com/graphhopper/matching/ViterbiMatchEntry.java:
```diff
@@ -0,0 +1,316 @@
+
+    /**
+     * DistanceCalc used for e.g. detecting how far candidates are from the original point.
+     * TODO: needs to be DistancePlanProject to be consistent with prior behavior - DistanceCalcEarth fails.
```
@karussell - are you aware that the current LocationIndexMatch uses DistancePlaneProjection and fails the tests with DistanceCalcEarth (which I assume is more accurate)?

karussell replied:
No, which tests exactly? I saw there were some minor modifications to them - why is it necessary to modify the tests?

kodonnell replied:
[This](https://github.com/graphhopper/map-matching/blob/master/matching-core/src/test/java/com/graphhopper/matching/LocationIndexMatchTest.java#L93-L95) one. The tests changes weren't associated with logic (and were e.g. API changes). A couple I refactored as testing "unknown 1 == unknown 2" does not imply "unknown 1 == correct && unknown 2 == correct".

karussell replied:
This is not a big problem here IMO. As the tests says: return at least 4-5, 4-7. So more precise means just less results here we should probably make the tests like set.contains() instead of equals to fix this .

kodonnell commented:
OK, finally got back to this. @karussell / @stefanholder, I think the map-matching logic/API is reasonably stable, and the only remaining work is the web stuff + doco + tests - which I'll get to if you both approve of the logic/API. Quick summary of some of the major changes (which should all be in code doco):

- wrap GPXEntry's as MatchEntry's to include match info. Allows one to easily see how the input GPX was used in the matching, and where it was matched to.
- MatchEntry's are wrapped as ViterbiMatchEntry's to allow their use with hmm-lib.
- MatchResult holds a sequence of MatchSequences which holds a sequence of MatchEdges.
- Temporal gaps between sequences are filled with 'UNKNOWN' sequences (as discussed above).
- LocationIndexMatch logic was moved to ViterbiMatchEntry.findCandidateLocations as it seemed to make more sense there. This requires LocationIndexTree.XFirstSearchCheck being public not protected.

kodonnell commented on matching-core/src/main/java/com/graphhopper/matching/MapMatching.java:
```diff
@@ -163,296 +155,279 @@ public void setMeasurementErrorSigma(double measurementErrorSigma) {
+                algoOptions.getWeighting().getFlagEncoder());
+        final List<List<QueryResult>> candidateLocationsPerEvent = new ArrayList<List<QueryResult>>();
+        for (ViterbiMatchEntry viterbiMatchEntry : viterbiMatchEntries) {
+            // TODO: shouldn't we find those within e.g. 5 * accuracy? Otherwise we're not
```
@karussell / @stefanholder - does that make sense? I haven't looked in details, but wouldn't the emission probability be non-zero for points outside measurementSigmaError? Not in scope of this PR, but thought it might be a bug.

stefanholder replied:
It probably makes more sense to use 2 * accuracy instead because then the GPS error would be within the candidate search radius in 95% instead of 68% of all cases assuming that the normal distribution models GPS noise correctly (see also [the 68–95–99.7 rule](https://en.wikipedia.org/wiki/68%E2%80%9395%E2%80%9399.7_rule)).

In addition to the (increased) search radius radius, I think it makes sense to only consider the k (e.g. 10) closest candidates around the GPS position. This also gives a more predictable map matching run time. The number of candidates and maybe also the search radius / sigma coefficient should be a user configurable parameter. I think this deserves a new issue.

> wouldn't the emission probability be non-zero for points outside measurementSigmaError

Yes. Since the emission probability is modeled as a normal distribution the emission probability is always greater than zero no matter how far away the candidate is from the GPS position. But for practical reasons we have to stop the search at some radius.

stefanholder commented:
>Removed the logging code as I didn't want to spend time refactoring that while I was developing. Can add it back in if desired.

I think that the logging code is desired, at least until we have ways to visualize candidates and matched paths. I strongly relied on the logging in PR #81 and PR #91 and I think that it will be useful for other map matching improvements. Maybe it's a good idea to put the logging code into separate methods to reduce the noise in the core code.

stefanholder commented:
Seems that you need to merge master into your branch because PR #91 has been pushed into master in the meantime.

kodonnell commented:
> logging code

I'm with you - it can make debugging easier, but it can also make the core harder to develop (at least with bigger changes like this). As a side note, useful logging is often specific to the feature/bug worked on at the time, and isn't generic. I don't know what the best approach is - I'm happy to add it in, or leave it out, or refactor it into other methods (though I'm not sure how to do that).

> need to merge master

Will do. I'd prefer to do the merge once I know this PR is likely to be included.

stefanholder commented:
>I'm happy to add it in, or leave it out, or refactor it into other methods (though I'm not sure how to do that).

I would then suggest that you add the logging in again and I will then refactor this into separate methods in another PR. This should also make future refactoring easier.

kodonnell commented:
Cool - logging added back in.

kodonnell commented:
Sorry, I was waiting on you both but had forgotten to merge master. How does it look?

karussell commented:
Is this still a WIP? I've not checked it closely yet, maybe @stefanholder did? The performance implications and also the renaming should be discussed (necessary? better names?)

I've also created this here and can fix if this is required https://github.com/graphhopper/graphhopper/issues/971

kodonnell commented:
> Is this still a WIP?

Yes. The actual map-matching logic is (nearly) ready for review, but there are still other tasks as per [here](https://github.com/graphhopper/map-matching/pull/87#issuecomment-272345734). I'm happy to polish off the rest if there's some indication this PR is 'good'.

> The performance implications ...

| key | current master branch | current sequences branch |
| --- | --- | --- |
|location_index_match.max | 6.660339 | 6.278257 |
|location_index_match.mean | 0.2038874168 | 0.217992223999999 |
|location_index_match.min | 0.001316 | 0.001381 |
|location_index_match.sum | 1019.437084 | 1089.96112 |
|map_match.max | 1554.185032 | 1651.807253 |
|map_match.mean | 242.08705704 | 243.51960916 |
|map_match.min | 3.57688 | 3.188165 |
|map_match.sum | 24208.705704 | 24351.960916 |
|measurement.count | 5000 | 5000 |
|measurement.seed | 123 | 123 |
|measurement.time | 32932 | 33115 |
|measurement.totalMB | 1524 | 1460 |
|measurement.usedMB | 42 | 35 |

Those are from the measurement class, but still not completely robust (e.g. might be off due to competing resource, so really should be run repeatedly etc.) but give a reasonable indication that nothing's really changed (which is as I'd expect).

> the renaming should be discussed (necessary? better names?)

Very open to feedback on that. I've tried to use names that represent their purpose, to make it easier for others to understand the code base.

stefanholder commented:
>I've not checked it closely yet, maybe @stefanholder did? 

No, I didn't look at it closely yet either.

stefanholder commented on matching-core/src/main/java/com/graphhopper/matching/MatchEdge.java:
```diff
@@ -0,0 +1,38 @@
+    public final long toTime;
+
+    public MatchEdge(EdgeIteratorState edge, long fromTime, long toTime) {
+        assert edge != null : "edge should not be null";
```
If a precondition of a public method is violated, an exception should rather be thrown (NullPointerException in this case).

kodonnell replied:
Thanks - may be a few more of those. Will look now.

stefanholder commented on matching-core/src/main/java/com/graphhopper/matching/MatchSequence.java:
```diff
@@ -0,0 +1,270 @@
+     * Time (inclusive, in milliseconds) when first began on this sequence. -1 if not set.
+     * TODO: is in inclusive?
+     */
+    private long fromTime = -1;
```
I would prefer using Long and setting this to null.

kodonnell replied:
Made it final instead of long - it should never be null or -1 now.

stefanholder replied:
OK, this is even better.

stefanholder commented on matching-core/src/main/java/com/graphhopper/matching/MatchSequence.java:
```diff
@@ -0,0 +1,270 @@
+     * Time (inclusive, in milliseconds) when last was on this sequence. -1 if not set.
+     * TODO: is in inclusive?
+     */
+    private long toTime = -1;
```
See above.

stefanholder commented on matching-core/src/main/java/com/graphhopper/matching/MatchEntry.java:
```diff
@@ -1,30 +1,80 @@
+        locked = true;
+    }
+    
+    protected void saveMatchingState(int sequenceIdx, int sequenceMatchEdgeIdx,
```
Why is this protected, do you plan to inherit from this class?

kodonnell replied:
The method is called externally (e.g. [here](https://github.com/graphhopper/map-matching/pull/87/files#diff-53888a7f87fdc3fba144b0c38b1a4364R273)) and I don't want it to be available to users - I could be missing something, but isn't this the purpose of 'protected'?

stefanholder replied:
In this case you should use [package-private](https://docs.oracle.com/javase/tutorial/java/javaOO/accesscontrol.html) (no modifier) instead of protected. Protected should only be used if you plan to access the method in derived classes.

kodonnell replied:
Sorry about that, and thanks for pointing it out.

stefanholder commented on matching-core/src/main/java/com/graphhopper/matching/MatchEntry.java:
```diff
@@ -1,30 +1,80 @@
+        this.gpxEntry = gpxEntry;
+    }
+    
+    protected void markAsNotUsedForMatching() {
```
Why is this protected, do you plan to inherit from this class?

kodonnell replied:
See above.

stefanholder commented on matching-core/src/main/java/com/graphhopper/matching/util/ViterbiMatchEntry.java:
```diff
@@ -67,7 +67,7 @@ public int compare(QueryResult o1, QueryResult o2) {
     };
 
-    public TimeStep(MatchEntry matchEntry) {
+    public ViterbiMatchEntry(MatchEntry matchEntry) {
```
`ViterbiMatchEntry` sounds like this class represents the output of the Viterbi matching but it is actually the input. What about `HmmTimeStep`?

kodonnell replied:
I can't remember now why I made that change = ) Anyway, I agree with you.

stefanholder commented on matching-web/src/main/java/com/graphhopper/matching/http/MatchServlet.java:
```diff
@@ -124,13 +124,13 @@ public void doPost(HttpServletRequest httpReq, HttpServletResponse httpRes)
+//                PathMerger pathMerger = new PathMerger().
+//                        setDouglasPeucker(peucker).
+//                        setSimplifyResponse(wayPointMaxDistance > 0);
+//                pathMerger.doWork(matchGHRsp, Collections.singletonList(path), tr);
```
Why was this commented out? If the code is really not used anymore it should be deleted.

kodonnell replied:
As I didn't know how to get it working - I'll update that when I get the web stuff working.

stefanholder commented on matching-web/src/test/java/com/graphhopper/matching/http/MatchResultToJsonTest.java:
```diff
@@ -37,74 +37,74 @@
-        Assert.assertEquals("wpts[1].x should exists", -38.9999, link.getJSONArray("wpts").getJSONObject(1).get("x"));
-
-    }
+//    protected List<MatchEdge> getEdgeMatch() {
```
What's the reason you gave up on this test?

kodonnell replied:
As above.

stefanholder commented on matching-core/src/main/java/com/graphhopper/matching/MatchEdge.java:
```diff
@@ -20,15 +20,33 @@
+ * 
  * @author kodonnell
  */
 public class MatchEdge {
```
How about MatchedEdge instead of MatchEdge?

kodonnell replied:
I think I was trying to be consistent with (the original) `MatchResult` as opposed to `MatchedResult` - happy to change all if you wish?

stefanholder replied:
I don't think that consistency is needed here, so I would be happy with `MatchedEdge` (the matched edge) and `MatchResult` (the result is not matched, so I think this is fine).

kodonnell replied:
Done

stefanholder commented on README.md:
```diff
@@ -1,125 +0,0 @@
-## Map Matching based on GraphHopper
```
README.md was removed by some mistake.

kodonnell replied:
Not sure what I did there ...

stefanholder commented on matching-core/src/main/java/com/graphhopper/matching/LocationIndexMatch.java:
```diff
@@ -1,153 +0,0 @@
- *
- * @author Peter Karich
- */
-public class LocationIndexMatch extends LocationIndexTree {
```
I would prefer to have this logic in a separate class since this is a feature that could also be used for other purposes besides map matching. I think this class should actually be moved to the Graphhopper core.

kodonnell replied:
I think it makes sense for the method `findCandidateLocations` to be part of `HmmTimeStep` as it's currently laid out - it felt cleaner to me (mainly because the code is only used for this one purpose). However, if Graphhopper had this index logic implemented, then yes, it'd be nicer, and the `findCandidateLocations` method would be much simpler. I'm not too worried though - if you want me to change it, I'm happy to.

stefanholder replied:
@karussell, do you think that LocationIndexMatch should be moved to the Graphhopper core?

karussell replied:
yes
(probably either with different name or separate class method)

kodonnell replied:
I'll look into it

stefanholder commented on matching-core/src/main/java/com/graphhopper/matching/MatchSequence.java:
```diff
@@ -0,0 +1,323 @@
+    /**
+     * The matched sequence, as returned from viterbi.computeMostLikelySequence().
+     */
+    public final List<SequenceState<Candidate, MatchEntry, Path>> matchedSequence;
```
I think the map matching result should not contain SequenceState objects since these include internal data that is not relevant for the user. Hence, the relevant information should be extracted from SequenceState into separate result objects.

kodonnell replied:
I could just make it private? The user information (edges, original points in sequence, etc.) can be obtained by the user without it.

stefanholder replied:
Sure, this is also fine.

kodonnell replied:
Had to be package-private as it's used in MapMatching.java for debugging.

stefanholder replied:
OK, no problem.

kodonnell commented:
> What about the remaining non-checked items?

They're yet to be completed. Are you happy with the sequence-specific stuff of this PR (logic and API)? If so, I'll work on those additional items.

stefanholder commented:
>Are you happy with the sequence-specific stuff of this PR (logic and API)?

I'm happy with it. But since the API has changed a lot, I think it's important that @karussell also has a look.

kodonnell commented:
Great. I'll review your other comments, and work on the remaining items. When that's done - yes, I think it'd be good to have @karussell review.

kodonnell commented:
Wow, getting the GUI to handle multiple paths was much more painful than I thought. Anyway:
<img width="824" alt="screen shot 2017-03-28 at 3 19 41 pm" src="https://cloud.githubusercontent.com/assets/8204904/24386073/021a5090-13ca-11e7-8639-a066ebe77b12.png">

That's using [this gpx](https://github.com/graphhopper/map-matching/files/874314/seq.txt). I also had to tweak locationIndexTree.findNClosest to make it not always return a result (within tiles but outside radius) - otherwise I couldn't break it into sequences. (See [here](https://github.com/graphhopper/graphhopper/issues/997#issuecomment-289643318).)

Yes!!

As mentioned, the GUI thus far is hacky (and not fully-functioning) - I almost think it'd be better to rewrite from scratch than to continue hacking around it. However, then I'd risk making breaking changes (i.e. features I don't understand) ... so I'm not sure.

The unit tests are probably more important - I discovered some bugs even just testing this.

We're starting to build our house, so this is becoming less of a priority for me. I might still do it, but my I shouldn't make promises I can't keep ...

karussell commented:
Cool, thanks a lot :) !

> We're starting to build our house, so this is becoming less of a priority for me

Sure :) !

INRIX-Trang-Nguyen commented:
Did these changes make into the latest code base? We're finding very few matches when inspecting GPXExensions of match result. We have hundreds of input GPX entries but GPXExtension contains in many cases only a single entry.

kodonnell commented:
@INRIX-Trang-Nguyen not that I know of. I'm closing this as I have no intention of actively working on it.
